### PR TITLE
Add a match(...).with_captures(...) matcher.

### PR DESCRIPTION
**Pull request by samphippen**
See #831 for details/discussion.

### CODE REVIEW

**Comment by samphippen** [comment]
So ruby 1.8.7 doesn't support named captures, does anyone have a thought on how we can avoid the syntax error here @rspec/rspec

**Comment by samphippen** [comment]
@rspec/rspec got it working on 1.8.7, can I get a review?

File: lib/rspec/matchers/built_in/captures.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        def description
+          return BuiltIn::Match.new(expected).description unless @expected_captures
+          "match with string #{@expected} have captures #{surface_descriptions_in(@expected_captures).inspect}"
+        end
+
+      private
+
+        def match(expected, actual)
+          return BuiltIn::Match.new(expected).matches?(actual) unless Regexp === actual
+          return BuiltIn::Match.new(expected).matches?(actual) unless @expected_captures
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/captures.rb]: Although minor perf wise, this is a bunch of extra instantiations of the same thing (3 across the class)

File: lib/rspec/matchers/built_in/captures.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        def initialize(match_data)
+          @match_data = match_data
+        end
+
+        # @api private
+        # Returns match data names for named captures
+        # @return Array
+        def names
+          match_data.names
+        rescue NameError
```
Comment (JonRowe) [inline lib/rspec/matchers/built_in/captures.rb]: So is this the part that was incompatible with 1.8.7?

Comment (samphippen) [inline lib/rspec/matchers/built_in/captures.rb]: This, and regular expressions cannot have named captures (so I have to use Regexp.new) instead of literals.

**Comment by samphippen** [comment]
@JonRowe is this good to merge?

File: lib/rspec/matchers/built_in/captures.rb (hunk 2 above)
Comment (myronmarston) [inline lib/rspec/matchers/built_in/captures.rb]: I'd prefer to see the 1.8.7-specific logic extracted into a conditionally defined version of the method. That makes it obvious what the logic is for and that it can be removed when we drop support for 1.8.7. As this currently exists that is non-obvious.

File: lib/rspec/matchers.rb [hunk 3]
```diff
@@ -698,7 +698,11 @@ def all(expected)
     #       matchers (due to how the custom matcher DSL was evaluated in 2.x,
     #       `match` could not be used there), but is no longer needed in 3.x.
     def match(expected)
-      BuiltIn::Match.new(expected)
+      if ::String === expected
+        BuiltIn::Captures.new(expected)
+      else
+        BuiltIn::Match.new(expected)
+      end
```
Comment (myronmarston) [inline lib/rspec/matchers.rb]: What about the inverse match? E.g.

expect("2015-01-01").to match(/(\d{4})-/).with_captures(...)

File: lib/rspec/matchers/built_in/captures.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
+module RSpec
+  module Matchers
+    module BuiltIn
+      # @api private
+      # Provides the regex with captures matcher
+      class Captures < BaseMatcher
```
Comment (jasonkarns) [inline lib/rspec/matchers/built_in/captures.rb]: I'm not usually one for inheritance, but is this a legit case for Captures to inherit from Match?

File: lib/rspec/matchers/built_in/captures.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            end
+          else
+            false
+          end
+        end
+      end
+
+      # @api private
+      # Used to wrap match data and make it reliable for 1.8.7
+      class ReliableMatchData
```
Comment (jasonkarns) [inline lib/rspec/matchers/built_in/captures.rb]: Suggestion: this looks like a prime use case for DelegateClass. `initialize` doesn't need defined, nor `captures` ?

Comment (JonRowe) [inline lib/rspec/matchers/built_in/captures.rb]: We don't use the standard library if we can avoid it as it causes false positives for other developers.

File: lib/rspec/matchers.rb (hunk 3 above)
Comment (JonRowe) [inline lib/rspec/matchers.rb]: Do you mean `to_not`?

File: lib/rspec/matchers/built_in/captures.rb (hunk 4 above)
Comment (JonRowe) [inline lib/rspec/matchers/built_in/captures.rb]: Actually I think that's a good improvement, it would eliminate the need to instantiate it at all, you could just call `super`

File: lib/rspec/matchers/built_in/captures.rb (hunk 2 above)
Comment (JonRowe) [inline lib/rspec/matchers/built_in/captures.rb]: :+1:

File: spec/rspec/matchers/built_in/captures_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  it "matches a regex without named captures" do
+    expect(/(a)/).to match("a123a").with_captures("a")
+  end
+
+  if RUBY_VERSION != "1.8.7"
+    it "matches a regex with named captures" do
+      expect(Regexp.new("(?<num>123)")).to match("a123a").with_captures(:num => "123")
+    end
+    it "matches a regex with a nested matcher" do
```
Comment (JonRowe) [inline spec/rspec/matchers/built_in/captures_spec.rb]: Needs an extra line break

File: lib/rspec/matchers.rb (hunk 3 above)
Comment (samphippen) [inline lib/rspec/matchers.rb]: no, this refers to the opposite case to the one now, where the regex is the target and the string is the value passed into the matcher.

Comment (myronmarston) [inline lib/rspec/matchers.rb]: Regardless, the point I was trying to make is that we should support with_captures for both regex matching against a strong and a string matching against a regex. All the specs I saw in this PR were only one way (the opposite of what I wrote here, I thought) and this logic looks like it only supports it one way.

Maybe it would be better to add a with_captures method to Match and have this always return a match instance? Besides, it seems odd to return the Captures instance if the user has not called with_captures and may not ever call that.

Comment (JonRowe) [inline lib/rspec/matchers.rb]: The other PR helped me understand that, should have said, sorry!

Comment (jasonkarns) [inline lib/rspec/matchers.rb]: @myronmarston adding with_captures to Match is the approach I took in #849. Two subclasses of Match are created (MatchWithCaptures, MatchWithNamedCaptures) and the appropriate one is instantiated/returned when with_captures is invoked. If it's never invoked, the matcher remains an instance of Match. The choice between MatchWithCaptures/MatchWithNamedCaptures is made based on whether the expected_captures are a Hash or hash-matcher.

**Comment by samphippen** [comment]
@JonRowe @myronmarston I've merged the code into the match matcher and I'm pretty pleased with how this is all looking, what do you think? Shall I squash, changelog and merge on green?

**Comment by jasonkarns** [comment]
I have a couple edge cases that I don't believe would be covered by this. Should I submit them as failing tests in a PR to this PR?

**Comment by samphippen** [comment]
@jasonkarns can you just provide a gist with failing examples?

**Comment by jasonkarns** [comment]
Sure, I'll get them up later today

**Comment by jasonkarns** [comment]
gisted: https://gist.github.com/jasonkarns/f009fa582315fae87cfb

**Comment by samphippen** [comment]
@myronmarston/@jonr do you have thoughts on this example:

```
it "matches indexed captures against named captures" do
  expect(Regexp.new("(?<num>123)")).to match("a123a").with_captures("123")
end
```

I could go either way with this passing.

**Comment by myronmarston** [comment]
I'd expect it to pass.

**Comment by JonRowe** [comment]
This LGTM, I'd merge once squashed.
